Adds article: "what is an orchestrator?" (#1707)
kestra-io/docs: content/blogs/2024-09-16-what-is-an-orchestrator.md

---
title: "How Orchestration Can Optimize Your Engineering Processes"
description: "Learn what an orchestrator is and why you should use it"
date: 2024-09-16T15:00:00
category: Solutions
author:
  name: Federico Trotta
  image: "ftrotta"
image: /blogs/2024-09-16-what-is-an-orchestrator.jpg
---

If you're an engineer looking to scale your automation - maybe because your company is growing rapidly — then this article is definitely for you.

Here, we’ll break down what an orchestrator is, why you might need one, and provide a practical example using Kestra.

Let’s dive in!

## What is an Orchestrator?

> In software engineering and data management, an orchestrator is a tool or platform that automates, manages, and coordinates various workflows and tasks across different services, systems, or applications.

Reviewer: ```suggestion
In software engineering and data management, an orchestrator is a tool that automates, manages, and coordinates various workflows and tasks across different services, systems, or applications.

```

either tool or platform is enough

> Think of it like a conductor of an orchestra, making sure all components perform in harmony, following a predefined sequence or set of rules. Whether you're dealing with data pipelines, microservices, or CI/CD systems, an orchestrator ensures everything runs smoothly, efficiently, and without manual intervention.

Reviewer: ```suggestion
Think of it like a conductor of an orchestra, making sure all components perform in harmony, following a predefined sequence or set of rules. Whether you're dealing with data pipelines, microservices, or CI/CD systems, an orchestrator ensures everything runs reliably without manual intervention.

```

## Orchestration vs. Automation 

> At this point, you might ask yourself: what's the difference between automation and orchestration? Well, these two concepts are somehow related, but are different. Let's see:

Reviewer: ```suggestion
What's the difference between automation and orchestration? These two concepts are related but not quite the same:

```

> - **Automation** refers to the execution of individual tasks or actions without manual intervention. For example, automatically triggering a test suite after a code commit is automation.

Reviewer: ```suggestion
- **Automation** refers to the execution of individual tasks or actions without manual intervention. For example, automatically triggering a test suite after a pull request is opened.

```

after a code commit is automation? You mean after a push to Git?

  
- **Orchestration** goes beyond automation by managing the flow of multiple interconnected tasks or processes. It defines not only what happens but also when and how things happen, ensuring that all tasks (whether automated or not) are executed in the correct order, with the right dependencies and error handling in place.

> In essence, while automation focuses on individual tasks, orchestration ensures all those tasks are arranged and managed within a broader, cohesive system. This is particularly vital in environments where complex processes with interdependent steps need to be handled efficiently.

Reviewer: ```suggestion
In essence, while automation focuses on individual tasks, orchestration ensures all those tasks are arranged and managed within a broader, cohesive system. This matters if you need to reliably handle complex processes with many interdependent steps.

```

### Examples of Orchestration

Now, let’s check out some orchestration examples:

- **Data pipeline orchestration**: Consider an ETL workflow where data is extracted from a source, transformed, and then loaded into a database. The orchestrator ensures these steps happen in sequence: first extracting the data, then transforming it, and finally loading it into the database. If one step fails, the orchestrator can retry or trigger an error-handling process.

- **CI/CD pipeline orchestration**: In a CI/CD pipeline, orchestration involves tasks like compiling code, running tests, deploying to a staging environment, and triggering manual approval for production deployment. The orchestrator ensures that each task runs in the correct order and only when the previous task has been successfully completed.

- **Cloud infrastructure orchestration**: When deploying a new environment in the cloud, an orchestrator manages the provisioning of servers, databases, and network configurations. It ensures that all resources are created in the right order, handling dependencies such as setting up the network before deploying a database.

## Benefits of Using an Orchestrator

> Today, IT teams manage a wide range of systems and applications across datacenters, clouds, and edge locations. As these environments grow in complexity, automating tasks can improve efficiency and make processes easier to manage. But as your company scales, you may find that only automation isn’t enough: this is where an orchestrator becomes essential.

Reviewer: ```suggestion
As IT environments become more complex, managing workflows manually becomes harder and prone to errors. An orchestrator simplifies this by offering a standardized way to schedule, run, and monitor workflows, making everything more predictable and manageable.

```

it might be best to ban the word efficiency from our vocabulary 😄

So, here are some key benefits of using an orchestrator:

> 1. **Efficiency and automation**: Orchestrators manage entire workflows, reducing the need for manual intervention. This leads to faster execution, fewer human errors, and improved consistency across tasks.

Reviewer: ```suggestion
1. **Faster time to value**: With a consistent way to schedule and run workflows, you avoid reinventing the wheel each time. This speeds up execution and helps your team focus on delivering outcomes faster.

```
efficiency is a loaded term

    commands:
      - python process_data.py

triggers:
  - id: schedule_trigger
    type: io.kestra.plugin.core.trigger.Schedule
    cron: 0 10 * * *

errors:
  - id: alert
    type: io.kestra.plugin.notifications.slack.SlackExecution
    channel: "#general"
    url: "{{ secret('SLACK_WEBHOOK') }}"
```

::alert{type="info"}
**Note**: The YAML defines the following:

- The `tutorial` namespace which is the subfolder where the Python file is stored.
- The type `io.kestra.plugin.scripts.python.Commands` is used to run Python files that are stored into Kestra. Read more [here](https://kestra.io/plugins/plugin-script-python/tasks/io.kestra.plugin.scripts.python.commands).
- `python process_data.py` executes the Python script.
- The `triggers` section adds the trigger. 
- The `errors` section manages eventual errors and sends a Slack message (you have to set up a dedicated Slack channel to make it work).

::

When you've done, click on **Execute**: you'll be asked to load the CSV file containing the data. When the job is done, in the logs section you'll see the results:

![results.png](/blogs/2024-09-16-what-is-an-orchestrator/result.png)

As you can see:

> - The YAML is shorter and simpler than the one used for GitHub actions.

Reviewer: ```suggestion
- The YAML is shorter and simpler than the one used for GitHub Actions.

```

- You can manage errors.

> - You don't need to create a repository in GH, as everything happens in Kestra's UI.

Reviewer: ```suggestion
- You don't need to create a repository in GitHub, as everything happens in Kestra's UI.

```

> Plus, Kestra provides a lot of features as, for examples, statistics on processes, pluginis that allow you to connest with your preferred tools, and more.

Reviewer: ```suggestion
Plus, Kestra provides hundreds of plugins that allow you to connect with your preferred tools.

```
